---
title: "Hdda2"
output: html_document
---

# IMPORT LIBRARIES

```{r setup, include=FALSE}

library("biomformat")
library("phyloseq")
library("microbiome")
library("vegan")
library("gplots")
library("gtools")
library("RVAideMemoire")
library("reshape2")
library("ggplot2")
library("corrplot")
library("tidyr")
library("ape")
library("splitstackshape")
library("readxl")
library("philentropy")
library("smacof")
library("ade4")
library("PCPS")
library("plotly")
library("heatmaply")
library("ggfortify")
library("arules")
```

# DATA IMPORT
```{r}
# Import OTU table and nutrients data

biom <- read_biom("datasets/221_otu_table.biom")
bacteria_taxa = observation_metadata(biom)
nutrients_name = read_xls("datasets/1208344TableS4.xls",sheet=1,skip = 1,col_names=TRUE)
colnames(nutrients_name) = nutrients_name[1,]
nutrients_name = nutrients_name[-1,]
colnames(nutrients_name) = c("nutrient_name","id")
nutrients_table = as.data.frame(read_xls("datasets/1208344TableS4.xls",sheet=2,skip = 0,col_names=TRUE))
colnames(nutrients_table)[1] = "id"

transposed_nutrients = as.data.frame(t(nutrients_table))
colnames(transposed_nutrients) = transposed_nutrients[1,]
transposed_nutrients = transposed_nutrients[-1,]
transposed_nutrients$id = rownames(transposed_nutrients)

nutrients_table_with_names = merge(transposed_nutrients,nutrients_name,by="id")
nutrients =as.data.frame(t(nutrients_table_with_names))
nutrients = nutrients[-1,]
colnames(nutrients) = nutrients_table_with_names$nutrient_name
map_nutrients = nutrients[-nrow(nutrients),]
map_nutrients$id = rownames(map_nutrients)
map_nutrients = as.data.frame(lapply(map_nutrients,function(x) as.numeric(as.character(x))))
rownames(map_nutrients) = map_nutrients$id

`%notin%` <- Negate(`%in%`)
map_nutrients = subset(map_nutrients, rownames(map_nutrients) %notin% "4017")
nutrients_table = subset(nutrients_table, id != 4017)



fecal_samples_mat = as(biom_data(biom), "matrix")

colnames(fecal_samples_mat) <- gsub("X","", colnames(fecal_samples_mat))
colnames(fecal_samples_mat) = unlist(lapply(colnames(fecal_samples_mat), function(x) strsplit(x,".C.")[[1]][2]))
colnames(fecal_samples_mat) = unlist(lapply(colnames(fecal_samples_mat), function(x) substr(x, start = 1, stop = 4)))






fecal_samples_data <- as.data.frame(t(fecal_samples_mat))
fecal_samples_data$id = rownames(fecal_samples_data)

ids = as.data.frame(map_nutrients$id)
names(ids) = "id"

fecal_samples_data = merge(ids,fecal_samples_data,by = "id")
fecal_samples_data = as.data.frame(t(fecal_samples_data))
colnames(fecal_samples_data) = fecal_samples_data[1,]
fecal_samples_data = fecal_samples_data[-1,]
fecal_samples_data$id = rownames(fecal_samples_data)

fecal_samples = as.matrix(fecal_samples_data[,-ncol(fecal_samples_data)])
```

```{r}
# Cleaning bacteria_taxa data (removed "[" and "]" in taxa names).
bacteria_taxa$taxonomy2 <- gsub("\\[|\\]", "", bacteria_taxa$taxonomy2)
bacteria_taxa$taxonomy3 <- gsub("\\[|\\]", "", bacteria_taxa$taxonomy3)
bacteria_taxa$taxonomy4 <- gsub("\\[|\\]", "", bacteria_taxa$taxonomy4)
bacteria_taxa$taxonomy5 <- gsub("\\[|\\]", "", bacteria_taxa$taxonomy5)
bacteria_taxa$taxonomy6 <- gsub("\\[|\\]", "", bacteria_taxa$taxonomy6)
```


```{r}
# Creation of phylogenetic tree starting from all 3393 bacteria.
fecal_samples_abundances <- apply(fecal_samples,2,function(x){x/sum(x)})
OTU <- otu_table(fecal_samples_abundances, taxa_are_rows = TRUE)
TAX <- tax_table(as.matrix(bacteria_taxa))
```

```{r}
physeq <- phyloseq(OTU, TAX)
random_tree = rtree(nrow(OTU), rooted=TRUE, tip.label=taxa_names(physeq))
phy_tax = merge_phyloseq(physeq, random_tree)


# Creation of phylogenertic tree starting from only 6th taxonomy level (genera).

physeq_genera <- microbiome::aggregate_taxa(physeq, level = "taxonomy6")


# Abundance plot
plot_bar(physeq,fill="taxonomy6") +
  ggtitle("Abbondanza relativa a livello di genere") +
     theme(axis.title.x=element_blank(),
           axis.text.x=element_blank(),
           ) +
  guides(fill=guide_legend(title="Tassonomia 6: generi")) +
  xlab("Soggetti in studio") +
  ylab("Abbondanza relativa") +
  theme(plot.title = element_text(hjust = 0.5))


random_tree_genera = rtree(nrow(tax_table(physeq_genera)), rooted=TRUE, tip.label=taxa_names(physeq_genera))
#plot(random_tree,show.tip.label = F)
phy_tax_genera = merge_phyloseq(physeq_genera, random_tree_genera)
```

# DATA PREPROCESSING

```{r}
# Divide taxonomy table into 5 different levels tables based on taxonomy rank:
# tax2 contains only kingdom and phylum and its name associated
# tax3 contains only kingdom, phylum, class and its name associated
# ... and so on until genera level

tax2 <- bacteria_taxa[,c("taxonomy1","taxonomy2")]
tax2["id"] <- rownames(tax2)
tax2["name"] <- tidyr::unite(tax2[,-ncol(tax2)],"merge_type",sep = ".")

tax3 <- bacteria_taxa[,c("taxonomy1","taxonomy2", "taxonomy3")]
tax3["id"] <- rownames(tax3)
tax3["name"] <- tidyr::unite(tax3[,-ncol(tax3)],"merge_type",sep = ".")

tax4 <- subset(bacteria_taxa[,c(1:4)], taxonomy4 != "o__")
tax4["id"] <- rownames(tax4)
tax4["name"] <- tidyr::unite(tax4[,-ncol(tax4)],"merge_type",sep = ".")

tax5 <- subset(bacteria_taxa[,c(1:5)], taxonomy5 != "f__")
tax5["id"] <- rownames(tax5)
tax5["name"] <- tidyr::unite(tax5[,-ncol(tax5)],"merge_type",sep = ".")

tax6 <- subset(bacteria_taxa[,c(1:6)], taxonomy6 != "g__")
tax6["id"] <- rownames(tax6)
tax6["name"] <- tidyr::unite(tax6[,-ncol(tax6)],"merge_type",sep = ".")
```

```{r}
# Delete taxonomies in each sub-dataframe to have only bacteria taxa names

taxa2 = tax2[,-c(1,2)]
taxa3 = tax3[,-c(1:3)]
taxa4 = tax4[,-c(1:4)]
taxa5 = tax5[,-c(1:5)]
taxa6 = tax6[,-c(1:6)]
```


```{r}
# Function to prepare data to be used. The function prepare_data takes in input a table that contains id and taxonomy name and matches names in fecal_samples_data to build a dataframe in wich rows are patients and columns are bacteria names based on selected taxonomy (taxa2 has kingdom and phylum, taxa3 kingdom, phylum and class, ...).

library(dplyr)
prepare_data <- function(taxonomy_table){
  fecal_samples_by_taxa <- as.data.frame(merge(fecal_samples_data,taxonomy_table,by="id", sort = F))
  fecal_samples_by_taxa$id <- NULL
  fecal_samples_by_taxa <- fecal_samples_by_taxa %>% group_by(name) %>% summarize_all(sum) 
  fecal_samples_by_taxa <- as.data.frame(t(fecal_samples_by_taxa))
  colnames(fecal_samples_by_taxa) <- as.character(unlist(fecal_samples_by_taxa[1,]))
  fecal_samples_by_taxa <- fecal_samples_by_taxa[-1,]
  return(fecal_samples_by_taxa)
}

# Application of the function prepare_data to taxa2,taxa3,...,taxa6 and concatenation of those dataframe to obtain a final dataframe with all possible taxonomies from the OTU table used.
# The final table has 100 individuals on rows and 198 columns representing each possible combination of bacteria taxa.


taxa_table_2 <- as.data.frame(prepare_data(taxa2))
taxa_table_3 <- as.data.frame(prepare_data(taxa3))
taxa_table_4 <- as.data.frame(prepare_data(taxa4))
taxa_table_5 <- as.data.frame(prepare_data(taxa5))
taxa_table_6 <- as.data.frame(prepare_data(taxa6))
taxa_table <- cbind(taxa_table_2, taxa_table_3, taxa_table_4, taxa_table_5, taxa_table_6)
```


```{r}

# Check if there are nutrients with same value for the entire column: they will be discarded because they are not usable for permanova. 

map_nutrients = map_nutrients[vapply(map_nutrients, function(x) length(unique(x)) > 1, logical(1L))]
```


 

```{r}

# Compute microbiome distance based on all possible 3393 bacteria in each patient microbiome. Used unweighted unifrac distance.
microbiome_distance_unweighted <- UniFrac(phy_tax, weighted=FALSE, normalized=TRUE, parallel=TRUE, fast=TRUE)

# Compute permanova on standardized nutrients to select only nutrients associated with gut microbiome at an FDR level 0f 25%.

set.seed(1234)
fdr_list <- c()
for (i in 1:(ncol(map_nutrients)-1)){
    permanova = adonis(microbiome_distance_unweighted ~ map_nutrients[,i],permutations = 299)
    fdr = unlist(permanova$aov.tab)[16][1]
    fdr_list = append(fdr_list,fdr)
}

n_nutrients = length(fdr_list[fdr_list <= 0.25])
selected_fdr_indexes = which(fdr_list <=.25)


# Final nutrients table according to FDR <= 0.25 criterion: 101 nutrients selected
final_nutrients = map_nutrients[,selected_fdr_indexes]
``` 

# SELECT TAXA

```{r}
# Extraction of all possible bacteria taxa from taxa_table which respect 2 criteria:
# - abundance >= 0.2%
# - present in at least 10% of patients (= 10 patients).

extract_taxa = function(final){
  final <- as.data.frame(t(final))
  final2 = as.data.frame(lapply(final,function(x) as.numeric(as.character(x))))
  rownames(final2) = rownames(final)
  final_ab <- as.data.frame(apply(final2,2, function(x){100*x/sum(x)}))
  ab_filter = apply(final_ab,1,function(x) length(which(x >=0.2))>=1)
  filtered_df = final_ab[ab_filter,]
  percent_filter = apply(filtered_df,1,function(x) length(which(x!=0))>10)
  selected_bacteria_taxa = filtered_df[percent_filter,]
  selected_bacteria = as.data.frame(t(selected_bacteria_taxa))
  rownames(selected_bacteria) = gsub("X","", rownames(selected_bacteria))
  return (selected_bacteria)
}

# extract 76 relevant taxa according to abundance threshold and presence in samples
selected_taxa = extract_taxa(taxa_table)
selected_taxa_names = colnames(selected_taxa)
```

# HEATMAP CORRELATION NUTRIENTS - TAXA

```{r}
 Instead of creating Heatmap with correlation between selected nutrients and all 76 selected taxa, the heatmap plot will be on Spearman correlation between selected nutrients and relevant genera taxa.

genera = extract_taxa(taxa_table_6)# only relevant genera 

genera_names = genera
# change colnames for genera df and nutrients df to make them readable
#correlation_data$Taxa = as.character(correlation_data$Taxa)
colnames(genera) = unlist(lapply(colnames(genera), function(x) strsplit(x,"f__")[[1]][2]))




colnames(genera) = gsub(".g__", " ", colnames(genera))

duplicated_genera = c("Ruminococcus","Prevotella")

extract_genera_names = function(name){
  if (!strsplit(name," ")[[1]][2] %in% duplicated_genera){
    name = strsplit(name," ")[[1]][2]
  }
  return (name)
}
colnames(genera) = lapply(colnames(genera), extract_genera_names)



microbiome_nutrients_data = merge(final_nutrients,genera,by=0,sort=FALSE)

rownames(microbiome_nutrients_data) = microbiome_nutrients_data[,1]
microbiome_nutrients_data[,c("Row.names")] <- list(NULL)
microbiome_nutrients_data[,c("id")] <- list(NULL)
a = cbind(final_nutrients_ab,genera)
nutrients_microbiome_correlations = cor(microbiome_nutrients_data,method= "spearman")[1:ncol(final_nutrients_ab),(ncol(final_nutrients_ab)+1):ncol(microbiome_nutrients_data)]

heatmaply(nutrients_microbiome_correlations,colors = rev(hcl.colors(100,palette="RdBu")), k_row = 4, k_col = 4,
          main ="Analisi correlazioni tra generi e nutrienti",fontsize_row = 8,column_text_angle = 90,key.title = "Correlazione di Spearman") %>% layout(height=1300,width=900) 

```

### CLUSTERING: 


# DISTANCE BETWEEN MICROBIOMES BASED ON 6th TAXA AGGREGATION
```{r}
# Compute Jensen Shannon, Euclidean and Bray-Curtis divergence matrices on gut microbiomes based on 119 taxa obtained.

otu_table_genera <- as.data.frame(t(otu_table(physeq_genera)))

dist_eucl = dist(otu_table_genera)
#dist_jsd = as.dist(philentropy::JSD(as.matrix(otu_table_genera)))
dist_jsd <- phyloseq::distance(physeq_genera, method = "jsd")
dist_bray = vegdist(otu_table_genera, method="bray")

# Compute PAM Partitioning Around Medoids clustering on distance matrix just obtained. Then evaluating the best number of clusters from 1 to 20 based on Silhouette Coefficient.
nclusters_eucl=c()
nclusters_jsd=c()
nclusters_bray=c()
for (k in 2:20) { 
			clu_eucl <- cluster::pam(dist_eucl,k,diss = T, metric = "euclidean")
			nclusters_eucl[k-1]= clu_eucl$silinfo$avg.width
			clu_jsd <- cluster::pam(dist_jsd,k,diss = T, metric = "euclidean")
			nclusters_jsd[k-1]= clu_jsd$silinfo$avg.width
			clu_bray <- cluster::pam(dist_bray,k,diss = T, metric = "euclidean")
			nclusters_bray[k-1]= clu_bray$silinfo$avg.width
	}
plot(nclusters_jsd, type="h", xlab="k clusters", ylab="Silhouette", main = "Valutazione numero di cluster con metodo JSD")

# Select best clustering method

print (c("optimal clusters numbers Bray-Curtis:",(which.max(nclusters_bray) + 1)))
print (c("Silhouette:",round(max(nclusters_bray),2)))
print (c("optimal clusters numbers JSD:",(which.max(nclusters_jsd) + 1)))
print (c("Silhouette:",round(max(nclusters_jsd),2)))
print (c("optimal clusters numbers Euclidean:",(which.max(nclusters_eucl) + 1)))
print (c("Silhouette:",round(max(nclusters_eucl),2)))

# JSD clustering has the higher Silhouette coefficient so it will be choosen.

```



```{r}
# Selected clustering method with 2 clusters for all types of distance used. 
selected_clustering <- cluster::pam(dist_jsd,2,diss = T, metric = "euclidean") #get cluster 
cluster_assignment = as.data.frame(selected_clustering$clustering)
colnames(cluster_assignment) = "cluster"


pcoa = dudi.pco(dist_jsd, full = TRUE)

# % of explained variance barplot
barplot(100*pcoa$eig/sum(pcoa$eig))

first_pcs = as.data.frame(cbind(pcoa$li[,1],pcoa$li[,2]))
colnames(first_pcs) = c("PC1","PC2")
rownames(first_pcs) = rownames(pcoa$li)

otu_table_generapc = otu_table_genera[vapply(otu_table_genera, function(x) length(unique(x)) > 1, logical(1L))]

data_pcs_bacteria = merge(first_pcs,otu_table_generapc,by=0)
data_pcs_bacteria$Row.names <- NULL

correlations_pc_bacteria = as.data.frame(cor(data_pcs_bacteria))

correlations_first_pc = correlations_pc_bacteria[with(correlations_pc_bacteria, order(-PC1)), ]
correlations_second_pc = correlations_pc_bacteria[with(correlations_pc_bacteria, order(-PC2)), ]

# Plot individuals in two cluster on 2 dimension from PCOA output

medoids = first_pcs[selected_clustering$medoids,]
medoids$group = cluster_assignment[selected_clustering$medoids,]
names(medoids) = c("x.medoid","y.medoid","cluster")

data_plot_pca_clu = merge(first_pcs,cluster_assignment,by=0)
data_plot_pca_clu$Row.names <- NULL
data_pca_clu = merge(data_plot_pca_clu,medoids,by="cluster")

ggplot(data_pca_clu, aes(x=PC1,y=PC2,color=factor(cluster)))+geom_point(size=2)+
  geom_point(aes(x=x.medoid,y=y.medoid),size=4,pch=18)+
  geom_segment(aes(x=x.medoid, y=y.medoid, xend=PC1, yend=PC2),size=0.6)+
  stat_ellipse(aes(x=PC1, y=PC2,color=factor(cluster), group=factor(cluster)),type = "norm") +
  ggtitle("PCoA sui generi dei batteri", subtitle = "PAM clustering sugli enterotipi basato su JSD") +
  scale_colour_discrete(name  ="cluster",
                            labels=c("1", "2")) +
  theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5)) +
  theme(plot.title=element_text(size = 28),legend.title=element_text(size=28), 
    legend.text=element_text(size=20),axis.title.x = element_text(size = 20),
  axis.text.x = element_text(size = 20),
  axis.title.y = element_text(size = 20),
  axis.text.y = element_text(size = 20))

```

```{r}
# Generating boxplot for two bacteria genera (Prevotella and Bacteroides) in the two enterotype clusters 

table_taxa_clusters = merge(otu_table_genera, cluster_assignment, by = 0)
rownames(table_taxa_clusters) = table_taxa_clusters$Row.names
table_taxa_clusters[,'Row.names'] <- list(NULL)

# preparing data for prevotella boxplot

bacteroides = table_taxa_clusters[,c("cluster","g__Bacteroides")]
colnames(bacteroides)[2] = "bacteroides"
bacteroides$cluster = as.factor(bacteroides$cluster)

prevotella_sum = table_taxa_clusters$k__Bacteria_p__Bacteroidetes_c__Bacteroidia_o__Bacteroidales_f__Prevotellaceae_g__Prevotella + table_taxa_clusters$k__Bacteria_p__Bacteroidetes_c__Bacteroidia_o__Bacteroidales_f__Paraprevotellaceae_g__Prevotella

prevotella = as.data.frame(cbind(table_taxa_clusters$cluster,prevotella_sum))
colnames(prevotella) = c("cluster","prevotella")
prevotella$cluster = as.factor(prevotella$cluster)


ruminococcus_sum = table_taxa_clusters$k__Bacteria_p__Firmicutes_c__Clostridia_o__Clostridiales_f__Ruminococcaceae_g__Ruminococcus + table_taxa_clusters$k__Bacteria_p__Firmicutes_c__Clostridia_o__Clostridiales_f__Lachnospiraceae_g__Ruminococcus

ruminococcus = as.data.frame(cbind(table_taxa_clusters$cluster,ruminococcus_sum))
colnames(ruminococcus) = c("cluster","ruminococcus")
ruminococcus$cluster = as.factor(ruminococcus$cluster)


# boxplot on Bacteroides abundance difference between groups
ggplot(bacteroides, aes(x=cluster, y=bacteroides, fill=cluster)) +
  geom_boxplot() +ggtitle("Abbondanza relativa Bacteroides") +
  theme(plot.title = element_text(hjust = 0.5))+
  theme(plot.title=element_text(size = 30),legend.title=element_text(size=28), 
    legend.text=element_text(size=28),axis.title.x = element_text(size = 28),
  axis.text.x = element_text(size = 28),
  axis.title.y = element_text(size = 28),
  axis.text.y = element_text(size = 28)
  )

ggplot(prevotella, aes(x=cluster, y=prevotella, fill=cluster)) +
  geom_boxplot() +ggtitle("Abbondanza relativa Prevotella") +
  theme(plot.title = element_text(hjust = 0.5))+
  theme(plot.title=element_text(size = 30),legend.title=element_text(size=28), 
    legend.text=element_text(size=28),axis.title.x = element_text(size = 28),
  axis.text.x = element_text(size = 28),
  axis.title.y = element_text(size = 28),
  axis.text.y = element_text(size = 28)
  )

ggplot(ruminococcus, aes(x=cluster, y=ruminococcus, fill=cluster)) +
  geom_boxplot() +ggtitle("Abbondanza relativa Ruminococcus") +
  theme(plot.title = element_text(hjust = 0.5))+
  theme(plot.title=element_text(size = 28),legend.title=element_text(size=28), 
    legend.text=element_text(size=28),axis.title.x = element_text(size = 28),
  axis.text.x = element_text(size = 28),
  axis.title.y = element_text(size = 28),
  axis.text.y = element_text(size = 28)
  )


```

# ASSOCIATIONS NUTRIENTS - ENTEROTYPE CLUSTERS

```{r}
table_taxa_clusters$id = rownames(table_taxa_clusters)
subject_clusters = table_taxa_clusters[,c("id","cluster")]

final_nutrients$id = rownames(final_nutrients)

# Merging cluster assignment and standardized nutrients based on patient id. Then divided subjects that belongs to specific cluster.
nutriens_with_clusters = merge(subject_clusters,final_nutrients,by="id")
nutriens_cl1 = nutriens_with_clusters[nutriens_with_clusters$cluster==1,]
nutriens_cl2 = nutriens_with_clusters[nutriens_with_clusters$cluster==2,]


# Compute strenght of associations between enterotype cluster and nutrients.
strength_nutrients_enterotype1_association = as.matrix(colMeans(nutriens_cl1[,-1]))
strength_nutrients_enterotype2_association = as.matrix(colMeans(nutriens_cl2[,-1]))
associations = cbind(strength_nutrients_enterotype1_association,strength_nutrients_enterotype2_association)

associations_data <- as.data.frame(associations)

associations_data <- associations_data[-1,]
rownames(associations_data) <- rownames(associations)[-1]
colnames(associations_data) = c("Enterotipo 1", "Enterotipo 2")


# Plot association between enterotype clusters and nutrients
heatmaply(as.data.frame(associations_data),colors = rev(hcl.colors(100,palette="RdBu")),
          main ="Associazione tra nutrienti ed enterotipi",fontsize_title = 40 ,fontsize_row = 13, fontsize_col = 8,column_text_angle = 90,dendrogram = "row",k_row = 4,key.title = "Associazione standardizzata",labcol="black") %>% layout(height=1600,width=1200)

```

# MERGING DATAFRAME WITH DEMOGRAPHICAL VARS AND BMI WITH NUTRIENTS (and/or bacteria taxa)

```{r}
#Unione df nutrienti selezionati dopo Permanova con il df che ha la variabile target bmi
# Merging selected nutrients data with demographical data 

final_nutrients$id <- gsub("1011.C.", "", rownames(final_nutrients))

final_nutrients$id <- gsub(".01.P\\d|.01.S\\d", "", final_nutrients$id)

bmi_df <- read_excel("datasets/bmi calculation with bmicategories.xlsx")
colnames(bmi_df)[1] <- "id"

bmi_nutrients <- merge(bmi_df, final_nutrients, by = "id", sort = F)
```

```{r}
# Creating unique dataframe with demographical variables, nutrients, taxa at 6th level (genera).

genera_taxa_abundances = otu_table_genera
genera_taxa_abundances$id <- gsub("1011.C.", "", rownames(genera_taxa_abundances))
genera_taxa_abundances$id <- gsub(".01.P\\d|.01.S\\d", "", genera_taxa_abundances$id)

# generate strings for cluster assignment
cluster_assignment$id = rownames(cluster_assignment)
cluster_assignment$id <- gsub("1011.C.", "", cluster_assignment$id)
cluster_assignment$id <- gsub(".01.P\\d|.01.S\\d", "", cluster_assignment$id)
cluster_assignment$cluster = ifelse(cluster_assignment$cluster==1,"Prevotella","Bacteroides")

dataset_demo_nutrients_taxa = merge(bmi_nutrients, genera_taxa_abundances, by = "id", sort = F)


pca_nutrients = as.data.frame(prcomp(final_nutrients[,-ncol(final_nutrients)],scale=TRUE)$x[,1:5])

#(100*(pca2$sdev^2)/sum(pca2$sdev^2)) # % of variance explained from each nutrients pc


colnames(pca_nutrients) = c("PC1_nut","PC2_nut","PC3_nut","PC4_nut","PC5_nut")



genera$id <- gsub("1011.C.", "", rownames(genera))
genera$id <- gsub(".01.P\\d|.01.S\\d", "", genera$id)
genera_without_ids = genera[,-ncol(genera)]
pca_genera = as.data.frame(prcomp(genera_without_ids)$x[,1:2])
colnames(pca_genera) = c("PC1_taxa","PC2_taxa")


#100*(pcoa$eig/sum(pcoa$eig)) # % of variance explained from each bacteria taxa (119) pc
pca_bacteria_taxa = pcoa$li[,1:6]
colnames(pca_bacteria_taxa) = c("PC1_taxa","PC2_taxa","PC3_taxa","PC4_taxa","PC5_taxa","PC6_taxa")
rownames(pca_bacteria_taxa) = rownames(otu_table_genera)

principal_components_df = merge(pca_nutrients,pca_bacteria_taxa,by=0)
colnames(principal_components_df)[1] = "id"
principal_components_df$id <- gsub("1011.C.", "", principal_components_df$id)
principal_components_df$id <- gsub(".01.P\\d|.01.S\\d", "", principal_components_df$id)

dataset_demo_nutrients_genera = merge(bmi_nutrients, genera, by = "id", sort = F)
principal_components_df_gen = merge(pca_nutrients, pca_genera, by = 0)
colnames(principal_components_df_gen)[1] = "id"
principal_components_df_gen$id <- gsub("1011.C.", "", principal_components_df_gen$id)
principal_components_df_gen$id <- gsub(".01.P\\d|.01.S\\d", "", principal_components_df_gen$id)

all_vars_dataset = merge(merge(dataset_demo_nutrients_taxa, principal_components_df, by = "id"),cluster_assignment, by ="id")

all_genera_dataset = merge(merge(dataset_demo_nutrients_genera, principal_components_df_gen, by = "id"),cluster_assignment,by="id")

#write.csv(all_vars_dataset, "datasets/all_vars_df.csv", row.names = F, quote = F)
#write.csv(all_genera_dataset, "datasets/all_genera_vars_df.csv", row.names = F, quote = F)
```



